### ISSUE TITLE
lib_build.py pulls files from freertos smp branch

### ISSUE THREAD

**Issue by stefaandesmet2003** [BOJIT/PlatformIO-FreeRTOS#1]
Hi,
I discovered your magic script in a platformio forum conversation. Love it!
But I have a compile error, because apparently the tasks.h file is pulled from the freeRTOS smp branch where vTaskSwitchContext has an additional parameter. See log below.
I managed to work around by adding -D FREERTOS_TAG=V10.4.5 in platformio.ini.
But I don't know if this solves the problem correctly, ie. does it now pull files only from the main branch?

By the way, is there a proper way to force execution of your lib_build.py ? On simply changing platformio.ini the script didn't execute, I had to remove the FreeRTOS-kernel subdirectory, and that forced a new execution of lib_build.py, and pulled the right files.
How can I see the logs from lib_build.py?

```

lib/FreeRTOS-platformio/FreeRTOS-Kernel/portable/GCC/ARM_CM3/port.c: In function 'xPortStartScheduler':
lib/FreeRTOS-platformio/FreeRTOS-Kernel/portable/GCC/ARM_CM3/port.c:350:5: error: too few arguments to function 'vTaskSwitchContext'
o     vTaskSwitchContext();
     ^~~~~~~~~~~~~~~~~~
In file included from lib/FreeRTOS-platformio/FreeRTOS-Kernel/portable/GCC/ARM_CM3/port.c:33:0:
lib/FreeRTOS-platformio/FreeRTOS-Kernel/include/task.h:3134:23: note: declared here
 portDONT_DISCARD void vTaskSwitchContext( BaseType_t xCoreID ) PRIVILEGED_FUNCTION;
```

**Comment by BOJIT** [comment]
Hi Stefaan,

Glad you like the library - I didn't realise that anyone else used it! do you have a link to the forum conversation?

In answer to the first question - adding `-D FREERTOS_TAG=V10.4.5` will solve the problem. Each tag points to a specific commit, so referencing a tag will get a specific commit, regardless of which branch it is on.
In the broader scheme of things, I will need to find out a way to deal with tags on the SMP branch. In the 15 year Git history of FreeRTOS this is the first time there has been a tagged commit that doesn't correspond to the main release.

I don't think I can purely filter based on branch, as FreeRTOS doesn't seem to have a consistent branching strategy for releases: see below.
![image](https://user-images.githubusercontent.com/35117353/139858687-ca895b94-6774-4a15-b67b-0438bfbfbef5.png)

For now I might just filter out tags that contain the string 'SMP' when getting the latest tag. That way, by default the non-SMP tag will be pulled normally, and if someone wants to use the SMP branch they can just override with the `platformio.ini` file.

As for the second question, the `lib_build.py` script should be called every time you build the code. I think this is deliberate, as dynamically pulling code from Git when building the library isn't really how PlatformIO is intended to be used.

The reason you had to delete the files to get the new version is actually a bug on my part: The build script checks to see if the cloned repo exists before cloning it again (Line 91: `kernel_path = os.path.realpath('./FreeRTOS-Kernel')
`). This is so that the repo doesn't get re-cloned every time you re-build. However, this doesn't account for if you change the referenced tag after building the library once. This should be a relatively simple fix: I'll try to put out a patch in the next week or so.

To see the logs just add standard `print()` statements to `lib_build.py`. They should appear near the top of the PlatformIO build logs.

I'd be interested to know if these 'dynamically built' libraries turn out to be practical. I'm planning on building wrappers for LwIP and libopencm3 at some point. But I'm still unsure how stable they are across different platforms.

Cheers,
James

**Comment by stefaandesmet2003** [comment]
Hi James,
no hurry for a fix, I'm a slow learner and only just starting on freertos.
here is the link mentioning your work :
https://community.platformio.org/t/how-to-add-source-directories-and-include-directories-for-libraries/21101/2
Might be interesting to get in touch with Max Gerhardt (one of the pio moderators) to have your approach integrated in platformio. Certainly useful for freertos, where support is rather limited now, and the examples in the freertos repo aren't easily built in platformio.
libopencm3 is nicely integrated with platformio I find, except for the fact that the framework-libopencm3 package is outdated; but can be simply replaced with a git clone from opencm3.

**Comment by BOJIT** [comment]
Hi, this should be fixed now - see release [2.1.3](https://platformio.org/lib/show/11145/PlatformIO-FreeRTOS).

Default behaviour is now to get the latest commit tag that does not contain the string "SMP". The script will check the local copy first, so the script should work offline if the local and requested tag match. Otherwise the local copy will always be replaced by the relevant tagged version.

**Comment by stefaandesmet2003** [comment]
Excellent. Thanks!
